src/PackageManagement/IDE/DeleteOnRestartManager.cs

```diff
@@ -0,0 +1,187 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.PackageManagement
+{
+    public class DeleteOnRestartManager : IDeleteOnRestartManager
+    {
+        // The file extension to add to the empty files which will be placed adjacent to partially uninstalled package
+        // directories marking them for removal the next time the solution is opened.
+        private const string DeletionMarkerSuffix = ".deleteme";
+        private const string DeletionMarkerFilter = "*" + DeletionMarkerSuffix;
+
+        private string _packagesFolderPath = null;
+
+        public event EventHandler<PackagesMarkedForDeletionEventArgs> PackagesMarkedForDeletionFound;
+
+        public DeleteOnRestartManager(string folderPath)
+        {
+            PackagesFolderPath = folderPath;
+        }
+
+        public DeleteOnRestartManager(
+            ISettings settings,
+            ISolutionManager solutionManager)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            if (solutionManager == null)
+            {
+                throw new ArgumentNullException(nameof(solutionManager));
+            }
+
+            Settings = settings;
+            SolutionManager = solutionManager;
+        }
+
+        public ISolutionManager SolutionManager { get; }
+
+        public ISettings Settings { get; }
+
+        public string PackagesFolderPath
+        {
+            get
+            {
+                // This can only be null when intialized during startup and the solution is not loaded at construction.
+                // Try to get the packages folder Path now.
+                if (_packagesFolderPath == null)
```

Comment:
Could the instance of IDeleteOnRestartManager be a singleton across a VS instance? If so, PackagesFolderPath could change between the opening of different solutions

Is that why the setter below is public? If so, I think we should just have 1 constructor on this, which only takes a PackagesFolderPath. Because, I don't think (know), if SolutionManager or Settings is used anywhere else in the class